File: content/problems/0947-most-stones-removed-with-same-row-or-column/solutions/stone-index-dsu/Solution.java
class Solution {
    public int removeStones(int[][] stones){
        int n=stones.length;
        int[] parent=new int[n];
        for(int i=0;i<n;i++)parent[i]=i;
        for(int i=0;i<n;i++)for(int j=i+1;j<n;j++)if(stones[i][0]==stones[j][0]||stones[i][1]==stones[j][1])union(i,j,parent);
        Set<Integer> roots=new HashSet<>();
        for(int i=0;i<n;i++)roots.add(find(i,parent));
        return n-roots.size();
    }
    private int find(int x,int[] p){
        return p[x]==x?x:(p[x]=find(p[x],p));
    }
    private void union(int a,int b,int[] p){
        a=find(a,p);
        b=find(b,p);
        if(a!=b)p[a]=b;
    }
}

File: content/problems/0947-most-stones-removed-with-same-row-or-column/solutions/stone-index-dsu/solution.cpp
class Solution {
    vector<int> parent;
    int find(int x){
        return parent[x]==x?x:parent[x]=find(parent[x]);
    }
    public:int removeStones(vector<vector<int>>& stones){
        int n=stones.size();
        parent.resize(n);
        iota(parent.begin(),parent.end(),0);
        for(int i=0;i<n;++i)for(int j=i+1;j<n;++j)if(stones[i][0]==stones[j][0]||stones[i][1]==stones[j][1]){
            int a=find(i),b=find(j);
            if(a!=b)parent[a]=b;
        }
        unordered_set<int> roots;
        for(int i=0;i<n;++i)roots.insert(find(i));
        return n-roots.size();
    }
};
